Connection with SSH

#### System information: 
- Archlinux
- DBeaver version 22.0.1

#### Connection specification:
- Postgresql, MySQL
- SSH tunel

#### Describe the problem you're observing:
While conncting to connection with configured ssh tunel I get error. 

![image](https://user-images.githubusercontent.com/23062169/161549761-1087936f-5c22-4e7e-a8f8-c56e00917ad8.png)


#### Steps to reproduce, if exist:

connect to ssh tunel connection

#### Include any warning/errors/backtraces from the logs

2022-04-04 16:09:12.005 - Connect with 'PostgreSQL - newprod - pg - db1  ssh' (postgres-jdbc-17dfb715719-5955c05824f912cf)
2022-04-04 16:09:12.007 - Adding identity key
2022-04-04 16:09:12.007 - Configure tunnel
2022-04-04 16:09:12.072 - Connection failed (postgres-jdbc-17dfb715719-5955c05824f912cf)
org.jkiss.dbeaver.model.exec.DBCException: Can't initialize tunnel
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:911)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:70)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.jkiss.dbeaver.DBException: Cannot establish tunnel to 91.199.55.223:1422
	at org.jkiss.dbeaver.model.net.ssh.SSHImplementationJsch.setupTunnel(SSHImplementationJsch.java:134)
	at org.jkiss.dbeaver.model.net.ssh.SSHImplementationAbstract.initTunnel(SSHImplementationAbstract.java:134)
	at org.jkiss.dbeaver.model.net.ssh.SSHTunnelImpl.initializeHandler(SSHTunnelImpl.java:76)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:906)
	... 3 more
Caused by: com.jcraft.jsch.JSchException: fromBase64: invalid base64 data
	at com.jcraft.jsch.Util.fromBase64(Util.java:63)
	at com.jcraft.jsch.KnownHosts.setKnownHosts(KnownHosts.java:219)
	at com.jcraft.jsch.KnownHosts.setKnownHosts(KnownHosts.java:55)
	at com.jcraft.jsch.JSch.setKnownHosts(JSch.java:316)
	at org.jkiss.dbeaver.model.net.ssh.SSHImplementationJsch.setupTunnel(SSHImplementationJsch.java:110)
	... 6 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 385 out of bounds for length 385
	at com.jcraft.jsch.Util.fromBase64(Util.java:51)
	... 10 more

--- a/plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/core/DesktopUI.java
+++ b/plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/core/DesktopUI.java
@@ -276,6 +276,11 @@ public class DesktopUI implements DBPPlatformUI {
     }
 
     @Override
+    public boolean confirmAction(String title, String message, boolean isWarning) {
+        return UIUtils.confirmAction(null, title, message, isWarning ? DBIcon.STATUS_WARNING : DBIcon.STATUS_QUESTION);
+    }
+
+    @Override
     public UserResponse showErrorStopRetryIgnore(String task, Throwable error, boolean queue) {
         return ExecutionQueueErrorJob.showError(task, error, queue);
     }

--- a/plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ui/DBPPlatformUI.java
+++ b/plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ui/DBPPlatformUI.java
@@ -61,6 +61,7 @@ public interface DBPPlatformUI {
     void showMessageBox(@NotNull final String title, @Nullable final String message, boolean error);
     void showWarningMessageBox(@NotNull final String title, @Nullable final String message);
     boolean confirmAction(String title, String message);
+    boolean confirmAction(String title, String message, boolean isWarning);
 
     UserResponse showErrorStopRetryIgnore(String task, Throwable error, boolean queue);
 

--- a/plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ui/console/ConsoleUserInterface.java
+++ b/plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/ui/console/ConsoleUserInterface.java
@@ -89,6 +89,11 @@ public class ConsoleUserInterface implements DBPPlatformUI {
     }
 
     @Override
+    public boolean confirmAction(String title, String message, boolean isWarning) {
+        return false;
+    }
+
+    @Override
     public UserResponse showErrorStopRetryIgnore(String task, Throwable error, boolean queue) {
         System.out.println(task);
         error.printStackTrace(System.out);
@@ -233,4 +238,5 @@ public class ConsoleUserInterface implements DBPPlatformUI {
 
         return false;
     }
+
 }

--- a/plugins/org.jkiss.dbeaver.net.ssh.jsch/src/org/jkiss/dbeaver/model/net/ssh/JSCHUIMessages.java
+++ b/plugins/org.jkiss.dbeaver.net.ssh.jsch/src/org/jkiss/dbeaver/model/net/ssh/JSCHUIMessages.java
@@ -1,0 +1,35 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2022 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.net.ssh;
+
+import org.eclipse.osgi.util.NLS;
+
+public class JSCHUIMessages extends NLS {
+	static final String BUNDLE_NAME = "org.jkiss.dbeaver.model.net.ssh.JSCHUIMessages"; //$NON-NLS-1$
+
+	public static String ssh_file_corrupted_dialog_title;
+	public static String ssh_file_corrupted_dialog_message;
+
+
+	static {
+		// initialize resource bundle
+		NLS.initializeMessages(BUNDLE_NAME, JSCHUIMessages.class);
+	}
+
+	private JSCHUIMessages() {
+	}
+}

--- a/plugins/org.jkiss.dbeaver.net.ssh.jsch/src/org/jkiss/dbeaver/model/net/ssh/JSCHUIMessages.properties
+++ b/plugins/org.jkiss.dbeaver.net.ssh.jsch/src/org/jkiss/dbeaver/model/net/ssh/JSCHUIMessages.properties
@@ -1,0 +1,2 @@
+ssh_file_corrupted_dialog_title = Host key verification failed
+ssh_file_corrupted_dialog_message = Can't verify host key due to corrupted records in known hosts file.\nYou can resolve this manually (recommended) or skip verification.\n\nDo you want to bypass host key verification? \nIt opens possibility of MAN-IN-THE-MIDDLE ATTACK!

--- a/plugins/org.jkiss.dbeaver.net.ssh.jsch/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationJsch.java
+++ b/plugins/org.jkiss.dbeaver.net.ssh.jsch/src/org/jkiss/dbeaver/model/net/ssh/SSHImplementationJsch.java
@@ -101,17 +101,7 @@ public class SSHImplementationJsch extends SSHImplementationAbstract {
                 }
 
                 session.setUserInfo(userInfo);
-
-                if (DBWorkbench.getPlatform().getApplication().isHeadlessMode() || configuration.getBooleanProperty(SSHConstants.PROP_BYPASS_HOST_VERIFICATION)) {
-                    session.setConfig("StrictHostKeyChecking", "no");
-                } else {
-                    File knownHosts = SSHUtils.getKnownSshHostsFileOrNull();
-                    if (knownHosts != null) {
-                        jsch.setKnownHosts(knownHosts.getAbsolutePath());
-                    }
-                    session.setConfig("StrictHostKeyChecking", "ask");
-                }
-
+                setupHostKeyVerification(session, configuration);
                 session.setServerAliveInterval(configuration.getIntProperty(SSHConstants.PROP_ALIVE_INTERVAL));
                 session.setTimeout(configuration.getIntProperty(SSHConstants.PROP_CONNECT_TIMEOUT));
 
@@ -135,6 +125,32 @@ public class SSHImplementationJsch extends SSHImplementationAbstract {
             }
 
             sessions[index] = session;
+        }
+    }
+
+    private void setupHostKeyVerification(Session session, DBWHandlerConfiguration configuration) throws JSchException {
+        if (DBWorkbench.getPlatform().getApplication().isHeadlessMode() ||
+            configuration.getBooleanProperty(SSHConstants.PROP_BYPASS_HOST_VERIFICATION)) {
+            session.setConfig("StrictHostKeyChecking", "no");
+        } else {
+            File knownHosts = SSHUtils.getKnownSshHostsFileOrNull();
+            if (knownHosts != null) {
+                try {
+                    jsch.setKnownHosts(knownHosts.getAbsolutePath());
+                    session.setConfig("StrictHostKeyChecking", "ask");
+                } catch (JSchException e) {
+                    if (e.getCause() instanceof ArrayIndexOutOfBoundsException) {
+                        if (DBWorkbench.getPlatformUI().confirmAction(JSCHUIMessages.ssh_file_corrupted_dialog_title, 
+                            JSCHUIMessages.ssh_file_corrupted_dialog_message, true)) {
+                            session.setConfig("StrictHostKeyChecking", "no");
+                        } else {
+                            throw e;
+                        }
+                    }
+                }
+            } else {
+                session.setConfig("StrictHostKeyChecking", "ask");
+            }
         }
     }
 

--- a/plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/KnownHostsVerifier.java
+++ b/plugins/org.jkiss.dbeaver.net.ssh.sshj/src/org/jkiss/dbeaver/model/net/ssh/KnownHostsVerifier.java
@@ -41,7 +41,7 @@ public class KnownHostsVerifier extends OpenSSHKnownHosts {
         KeyType type = KeyType.fromKey(key);
 
         boolean isConfirmed = platformUI.confirmAction(SSHJUIMessages.verify_connection_confirmation_title,
-            NLS.bind(SSHJUIMessages.verify_connection_confirmation_message, new String[]{hostname, type.toString(), SecurityUtils.getFingerprint(key)}));
+            NLS.bind(SSHJUIMessages.verify_connection_confirmation_message, new String[]{hostname, type.toString(), SecurityUtils.getFingerprint(key)}), true);
 
         if (!isConfirmed) {
             return false;

